### ISSUE TITLE
[Addon development] Ability to provide torrent files as stream sources

### ISSUE THREAD

**Issue by detarkende** [Stremio/stremio-features#676]
**Stremio Version**
All versions.

**Is your feature request related to a problem? Please describe.**
[Disclaimer]: I'm new to the bittorrent protocol and this whole world, so I might have misunderstood some things. Please feel free to correct me, I'm eager to learn.
I'm building an addon for a private torrent site. As far as I understood, this means that the user's passkey is included in the tracker urls. This means that just the info hash is not enough to download the files.

**Describe the solution you'd like**
I'm not sure what bittorrent client Stremio uses internally, but in most cases, it's possible to provide torrent files or URLs pointing to torrent files to the client. If this is the case, I would like to provide the url to the torrent file in my stream handler.

**Describe alternatives you've considered**
- ability to provide a magnet url (afaik, magnet urls can specify trackers).
- ability to provide the trackers from the torrent file.
- ability to provide the torrent file itself as a buffer maybe?

**Additional context**
I might have misunderstood some things, but I have built the addon itself and noticed that the download doesn't start. So I tried to convert the info hash into a magnet url, and downloading from that on my local transmission instance, and it seems like that doesn't work either. Based on this, I'm assuming that the client can't find the trackers (since the tracker's url has to include the passkey).

**Comment by jaruba** [comment]
so you say you are building an addon, and want to reply with a torrent stream for a specific infohash, a specific file index and specific trackers

look into `infoHash` and `fileIdx` from the stream response: https://github.com/Stremio/stremio-addon-sdk/blob/master/docs/api/responses/stream.md

to set the trackers, you need to set the `sources` as an array of strings in the stream response, each tracker prefixed with `tracker:` see here: https://github.com/BoredLama/stremio-jackett-addon/blob/b55feb4349dc383f14c5903015e5eeb90836bdd3/index.js#L72
wether u choose to include the `dht:` too in the tracker list is ur choice, it is good for public torrents, but there are some private trackers that explicitly prohibit using the dht (because it makes the torrents public)

u should parse the magnet link to convert it to a stream response as per the links i provided, Stremio does not support torrent files or magnet links from addons, it only supports the streams object response as per the addons sdk docs

this also means that torrents that require a torrent file may not work with Stremio, to explain the issue behind this, u should understand that all torrent clients require a torrent file to start the torrent, magnet links are just a method of retrieving the .torrent file from the torrent swarm (instead of a site or other methods), the torrent file is required because it includes data such as the list of files from the torrent, and the torrent piece size (which cannot be included in magnet links)

the stream object response from addons works similarly to magnet links, and although Stremio will not support a typical link to a torrent file, it will still attempt to download the torrent file from the swarm (as it cannot start the torrent otherwise), some private trackers choose to not give magnet links, and only give torrent files, the torrent files that they give can include a flag stating that it should not be shared to the swarm, so in this case Stremio will fail to retrieve the torrent file for it from the swarm

to summarize:
- u can set infohash, file index and trackers in stream objects
- Stremio does not support links to torrent files in stream object in any way
- u can convert a torrent file to a magnet link (kind of), a magnet link to a stream object, a torrent file to a stream object, but when using private trackers it is possible that torrent file conversion to anything else will fail to start the torrent (if the flag to not share the torrent file was set)

**Comment by detarkende** [comment]
@jaruba Thank you for this very detailed answer. 
After some investigation, I believe the problem I was running into is that I was trying to add torrents from private sites as magnet urls, which doesn't work.
I changed my approach, so now I download the torrents on the addon server, with WebTorrent, then stream the results to the client. I'm still running into some issues connected to Range requests, but that's completely separate from this issue.

Again, thank you for your great help, you guys are the best 🚀
